packages/babel-parser/src/tokenizer/index.js

```diff
@@ -790,6 +790,40 @@ export default class Tokenizer extends ParserErrors {
     }
   }
 
+  readToken_at(): void {
+    // Increase past @
+    ++this.state.pos;
+    const next = this.input.charCodeAt(this.state.pos);
+
+    if (isIdentifierStart(next)) {
+      // Read the next word with a lookahead state that we'll
+      // throw away afterwards. Cheaper than doing a full lookahead,
+      // since we know that we're trying to read a word/identifier.
+      const old = this.state;
+      this.state = this.createLookaheadState(old);
+
+      const word = this.readWord1(next);
+      const wordNext = this.input.charCodeAt(this.state.pos);
+      const containsEsc = this.state.containsEsc;
+
+      // Replace the previous state
+      this.state = old;
+
+      if (wordNext === charCodes.colon) {
```

Comment:
We need to check the next character after `:` because it may form a doubleColon token `::`, e.g.

```js
class C {
  @init::b p;
}
```
is [valid](https://babeljs.io/repl#?browsers=safari%209&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1oAEAmAuLAjaADgNxIC-QA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=stage-0%2Cstage-1%2Ctypescript&prettier=false&targets=&version=7.16.2&externalPlugins=&assumptions=%7B%7D) given that `decorators-legacy` and `functionBind` enabled.